### PR TITLE
add an option to export csv file with verbose column names

### PR DESCRIPTION
**Pull request by xcaptain**
### CATEGORY

Choose one

- [ ] Bug Fix
- [x] Enhancement (new features, refinement)
- [ ] Refactor
- [ ] Add tests
- [ ] Build / Development Environment
- [ ] Documentation

### SUMMARY
In some cases we need to export csv file with custom table headers, this change enables to download csv with labels configured in datasource columns.

### TEST PLAN
1. in chart detail, edit datasource and add a label for each column
2. click the 'export to .csv format' button to export csv
3. the exported csv file will contain those label in first row

### CODE REVIEW

File: superset/viz.py [hunk 1]
```diff
@@ -491,6 +491,10 @@ def data(self):
     def get_csv(self):
         df = self.get_df()
         include_index = not isinstance(df.index, pd.RangeIndex)
+        if config.get("CSV_EXPORT_VERBOSE") == True:
```
Comment (villebro) [inline superset/viz.py]: This tripped `flake8`, can be expressed
```python
if config.get("CSV_EXPORT_VERBOSE"):
```

File: superset/config.py [hunk 2]
```diff
@@ -258,6 +258,7 @@
 # method.
 # note: index option should not be overridden
 CSV_EXPORT = {"encoding": "utf-8"}
+CSV_EXPORT_VERBOSE = True
```
Comment (villebro) [inline superset/config.py]: This could be expressed more verbosely (pun unintended), like `CSV_EXPORT_USE_VERBOSE_NAMES_AS_HEADERS` or similar.

**Review by villebro — CHANGES_REQUESTED** [review]

**Comment by xcaptain** [comment]
@villebro Thank you for the review, I fixed the naming issue and rebased the pull request. this enhancement works well on my server, hope it can get merged, it's really useful for those none-tech people to export a meaningful CSV file.

**Comment by villebro** [comment]
@xcaptain you need to run `black` on the file you've edited as it is now a requirement on CI.

File: superset/config.py [hunk 3]
```diff
@@ -258,6 +258,7 @@
 # method.
 # note: index option should not be overridden
 CSV_EXPORT = {"encoding": "utf-8"}
+CSV_EXPORT_USE_VERBOSE_NAMES_AS_HEADERS = True
```
Comment (villebro) [inline superset/config.py]: This should not default to `True`, as people may be depending on the current behaviour.

**Review by villebro — CHANGES_REQUESTED** [review]
Small change requested.

**Comment by xcaptain** [comment]
image
image

Added some images to show how this change work

**Review by villebro — APPROVED** [review]
LGTM. It would be neat to have a few unit tests for this (for both `True` and `False` case), but I think we can live without them if that feels difficult (I can add them later if needed).

**Comment by xcaptain** [comment]
Thanks, actually I don't know how to write unit tests for this commit, would be awesome if you could help writing tests. Do I need to rebase before it gets merged?

File: superset/viz.py [hunk 4]
```diff
@@ -491,6 +491,18 @@ def data(self):
     def get_csv(self):
         df = self.get_df()
         include_index = not isinstance(df.index, pd.RangeIndex)
+        if config.get("CSV_EXPORT_USE_VERBOSE_NAMES_AS_HEADERS"):
+            d = {
```
Comment (villebro) [inline superset/viz.py]: This dict should probably be called something more descriptive, e.g. `columns_with_verbose_names`

File: superset/viz.py [hunk 5]
```diff
@@ -491,6 +491,18 @@ def data(self):
...
+                o.column_name: o.verbose_name
+                for o in self.datasource.columns
+                if hasattr(o, "verbose_name")
```
Comment (villebro) [inline superset/viz.py]: I'm not sure how these objects usually look, but what if the attr is present, but is `None` or `''`? I would probably add a check to see that it's defined, aka. `if hasattr(o, "verbose_name") and o.verbose:` Do you agree? also, `o` could probably be called `column`.

File: superset/viz.py [hunk 6]
```diff
@@ -491,6 +491,18 @@ def data(self):
...
+            }
+            replaced_columns = {
+                column_name: d.get(column_name)
+                for column_name in df.columns
+                if d.get(column_name)
+            }
```
Comment (villebro) [inline superset/viz.py]: Is there some reason this is done in two steps? I wouöd think that it should be adequate to do dict comprehension on the `dd` columns and then just checking if those cols have a verbose name?

**Review by villebro — CHANGES_REQUESTED** [review]
Sorry for picking up on these late, but a few additional questions/change requests. I can fill in unit tests later, so don't worry about those.

File: superset/viz.py (hunk 4 above)
Comment (xcaptain) [reply superset/viz.py]: Thanks, fixed.

File: superset/viz.py (hunk 5 above)
Comment (xcaptain) [reply superset/viz.py]: image
Yes you are right, sometimes `verbose_name` would be `None`.

File: superset/viz.py (hunk 6 above)
Comment (xcaptain) [reply superset/viz.py]: Fixed, please take another look.

Comment (villebro) [reply superset/viz.py]: What I was thinking was that as there is already a loop of all columns in `self.datasource.columns`, one could just add an `and`to the if of the first dict comprehension to check whether or not that column is already present in the dataframe. That would save the effort of doing yet another dict comprehension. Then the if part would look like this:
```python3
                if hasattr(column, "verbose_name") and column.verbose_name and column.column_name in df.columns
```
which would make it possible to skip the second dict comprehension and just do
```python3
            df = df.rename(columns=columns_with_verbose_names)
```

Comment (xcaptain) [reply superset/viz.py]: Thank you for the guidance, your solution is indeed much clear

**Review by villebro — APPROVED** [review]
LGTM

**Comment by mistercrunch** [comment]
This seems like a setting that would apply to each export (as opposed to a global configuration setting...)

**Comment by xcaptain** [comment]
@villebro I rebased my commits, it's now ready to be merged. @mistercrunch it's useful to use verbose names as csv headers because usually only non-tech people read those csv files. but json files are usually used by programmer, so I think it's enough to only add this option for csv exporting.

**Comment by villebro** [comment]
@mistercrunch were you proposing adding a sub-menu under the export button or something similar?

**Comment by xcaptain** [comment]
ping @villebro , can this commit get merged?

**Comment by mistercrunch** [comment]
Yes I meant something like that. It's unclear what people may want, or even whether they care, and a global config setting doesn't seem like the right way to go about it.

Either we always prefer verbose names, or give the option to users on export. The option itself may be confusing to most users...

**Comment by villebro** [comment]
In that case I vote always using verbose names. Quickly toyed with the idea of having a submenu, but that felt overkill and confusing to users who just want to get the data out.

**Pull request closed without merging**